x-pack/plugins/security/server/lib/authentication/providers/token.js

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { canRedirectRequest } from '../../can_redirect_request';
+import { AuthenticationResult } from '../authentication_result';
+import { DeauthenticationResult } from '../deauthentication_result';
+
+/**
+ * Object that represents available provider options.
+ * @typedef {{
+ *  protocol: string,
+ *  hostname: string,
+ *  port: string,
+ *  basePath: string,
+ *  client: Client,
+ *  log: Function
+ * }} ProviderOptions
+ */
+
+/**
+ * Checks the error returned by Elasticsearch as the result of `authenticate` call and returns `true` if request
+ * has been rejected because of expired token, otherwise returns `false`.
+ * @param {Object} err Error returned from Elasticsearch.
+ * @returns {boolean}
+ */
+function isAccessTokenExpiredError(err) {
+  return err.body
+    && err.body.error
+    && err.body.error.reason === 'token expired';
+}
+
+/**
+ * Provider that supports request authentication via Basic HTTP Authentication.
+ */
+export class TokenAuthenticationProvider {
+  /**
+   * Server options that may be needed by authentication provider.
+   * @type {?ProviderOptions}
+   * @protected
+   */
+  _options = null;
+
+  /**
+   * Instantiates BasicAuthenticationProvider.
+   * @param {ProviderOptions} options Provider options object.
+   */
+  constructor(options) {
+    this._options = options;
+  }
+
+  /**
+   * Performs request authentication using Basic HTTP Authentication.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @param {Object} [state] Optional state object associated with the provider.
+   * @returns {Promise.<AuthenticationResult>}
+   */
+  async authenticate(request, state) {
+    this._options.log(['debug', 'security', 'token'], `Trying to authenticate user request to ${request.url.path}.`);
+
+    // first try from login payload
+    let authenticationResult = await this._authenticateViaLoginAttempt(request);
+
+    // if there isn't a payload, try basic auth
+    if (authenticationResult.notHandled()) {
+      authenticationResult = await this._authenticateViaHeader(request);
+    }
+
+    // if we still can't attempt auth, try authenticating via state (session token)
+    if (authenticationResult.notHandled()) {
+      authenticationResult = await this._authenticateViaState(request, state);
+      if (authenticationResult.failed() && isAccessTokenExpiredError(authenticationResult.error)) {
+        authenticationResult = this._authenticateViaRefreshToken(request, state);
```

Comment:
I'm not seeing the `_authenticateViaRefreshToken` implementation yet. It's likely this is known and intentional, but I wanted to note it just in case it wasn't.

Reply:
Ah yeah, I meant to call that out. It's not implemented yet, but it's essentially a copy+paste job from saml.